### ISSUE TITLE
TypeScript should warn about return statement inside generator functions

### ISSUE THREAD

**Issue by tmkn** [microsoft/TypeScript#32235]
**TypeScript Version:**  3.0.0-dev.201xxxxx

**Search Terms:**  yield return generator

Currently it is possible to have the following code:
**Code**

```ts
async function *myGeneratorFunction(): AsyncIterableIterator<number> {
    return 1;
    return 2;
    return 3;
}
```

**Expected behavior:**
A generator function that yields numbers 1 - 3

**Actual behavior:**
No syntax/runtime error but due to a human mistake the generator function is useless.

I copy/pasted an existing generator function, deleted the body with the new implementation and mistakenly used `return` instead of `yield`.

It would be nice if TypeScript could warn when it's encountering a `return` with an actual value inside a generator function.
It makes no sense to `return` a value.
Could be combined with a codefix in the IDE: `Did you mean yield?`
**Playground Link:**  

**Related Issues:**

**Comment by sandersn** [comment]
`return` is legal in generators because they are more powerful than iterators. However, a did-you-mean-yield codefix is a good idea when we encounter *dead* return statements.

**Comment by fatcerberus** [comment]
> It would be nice if TypeScript could warn when it's encountering a return with an actual value inside a generator function.
> It makes no sense to return a value.

Not 100% true - the `return`'d value is accessible:

```ts
function* g() { return 812; }
let iter = g();
console.log(iter.next());  // { done: true, value: 812 }
```

However, *dead* returns are indeed likely an indicator that `yield` was intended.

**Comment by fatcerberus** [comment]
Now that I think about it, maybe *any* dead code in a generator could trigger the codefix?  Say if you had `return` inside of a loop where you meant `yield`, the rest of the loop and function might be marked as dead code.

**Comment by tmkn** [comment]
> Not 100% true - the `return`'d value is accessible:
> 
> ```ts
> function* g() { return 812; }
> let iter = g();
> console.log(iter.next());  // { done: true, value: 812 }
> ```
Interesting, because I tested exactly this before creating the ticket, this is what I get in Google Chrome
```
function* g() { return 812; }
undefined
for(let foo of g()) {console.log(foo)}
undefined
```
In case of `for of` the `returned` value doesn't make it into the loop.
That's why assumed that `returning` any value is of no use.
Idk about the specs, maybe it's an error in Chrome.

**Comment by fatcerberus** [comment]
That's the expected behavior, the value returned when `{ done: true }` is not iterated by `for...of`, but you can access it if you call the iterator's `.next()` method directly.

**Comment by tmkn** [comment]
Thanks for the clarification.

Is there any use case where you want to `return` a value from a generator function, cause as I understand `done: true` will be automatically set and thus if consumed via `for of` you'll never get the value.

I still think that any non void `return` inside a generator function should be treated as an error as most likely `yield` was meant instead. a void `return` is fine as it can be used to reset the generator

In any case I would like to try to fix this myself, any starting points would be appreciated :)
For now I can check for dead code.

**Comment by sandersn** [comment]
@tkmn Look in src/services/codefixes/fixUnreachableCode.ts after making sure that the dead code error works properly in generators (it did in my quick testing). It's a pretty simple codefix that replaces statements with empty blocks. Probably the hardest part is finding the containing function of a particular token. I think we have some utilities in src/compiler/utilities.ts that can help.

**Comment by fatcerberus** [comment]
@tmkn I would say this issue https://github.com/microsoft/TypeScript/pull/30790 is a witness to the fact that the full power of generators is useful to people beyond the baseline use case as simple iterators for `for..of`.
